dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/TupleSet.java

```diff
@@ -0,0 +1,230 @@
+package com.dat3m.dartagnan.wmm.utils;
+
+import com.dat3m.dartagnan.program.event.core.Event;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.stream.Collectors;
+
+public class TupleSet {
+
+    private final Map<Event, Set<Event>> data;
+
+    public TupleSet() {
+        this.data = new HashMap<>();
+    }
+
+    public TupleSet(TupleSet other) {
+        this.data = new HashMap<>();
+        other.data.forEach((k, v) -> this.data.put(k, new HashSet<>(v)));
+    }
+
+    protected TupleSet(Map<Event, Set<Event>> data) {
+        this.data = data;
+    }
+
+    public boolean isEmpty() {
+        return data.isEmpty();
+    }
+
+    public int size() {
+        return data.values().stream()
+                .map(Set::size)
+                .reduce(0, Integer::sum);
+    }
+
+    public boolean contains(Event e1, Event e2) {
+        return data.containsKey(e1) && data.get(e1).contains(e2);
+    }
```

Comment:
This code makes one redundant lookup. I don't think the code is used in a hot path, so feel free to keep it as is.

Reply:
Fixed. These is still a room for optimizations, both in the TupleSet class and in the classes using it, but I don't want to spend time on this now as it is not the bottleneck.